src/HttpClientFactory/Http/test/DependencyInjection/HttpClientFactoryServiceCollectionExtensionsTest.cs

```diff
@@ -996,5 +1020,149 @@ public TypedClientWithTransientService(HttpClient httpClient, TransientService s
 
             public TransientService Service { get; }
         }
+
+        // https://github.com/nunit/nunit/blob/master/src/NUnitFramework/framework/Internal/SingleThreadedSynchronizationContext.cs
+        private class SingleThreadedSynchronizationContext : SynchronizationContext, IDisposable
+        {
+            private readonly Queue<ScheduledWork> _queue = new Queue<ScheduledWork>();
+            private Status status;
+
+            private enum Status
+            {
+                NotStarted,
+                Running,
+                ShutDown
+            }
+
+            public static void Run(Action action)
+            {
+                var previous = SynchronizationContext.Current;
+                var context = new SingleThreadedSynchronizationContext();
+                SynchronizationContext.SetSynchronizationContext(context);
+                try
+                {
+                    action();
+                }
+                finally
+                {
+                    SynchronizationContext.SetSynchronizationContext(previous);
+                }
+            }
+
+            public override void Post(SendOrPostCallback d, object state)
+            {
+                AddWork(new ScheduledWork(d, state, finished: null));
+            }
+
+            public override void Send(SendOrPostCallback d, object state)
+            {
+                if (SynchronizationContext.Current == this)
+                {
+                    d.Invoke(state);
+                }
+                else
+                {
+                    using (var finished = new ManualResetEventSlim())
+                    {
+                        AddWork(new ScheduledWork(d, state, finished));
+                        finished.Wait();
+                    }
+                }
+            }
+
+            private void AddWork(ScheduledWork work)
+            {
+                lock (_queue)
+                {
+                    if (status == Status.ShutDown) throw CreateInvalidWhenShutDownException();
+                    _queue.Enqueue(work);
+                    Monitor.Pulse(_queue);
+                }
+            }
+
+            public void ShutDown()
+            {
+                lock (_queue)
+                {
+                    status = Status.ShutDown;
+                    Monitor.Pulse(_queue);
+
+                    if (_queue.Count != 0)
+                        throw new InvalidOperationException("Shutting down SingleThreadedTestSynchronizationContext with work still in the queue.");
+                }
+            }
+
+            private static InvalidOperationException CreateInvalidWhenShutDownException()
+            {
+                return new InvalidOperationException("This SingleThreadedTestSynchronizationContext has been shut down.");
+            }
+
+            public void Run()
```

Comment:
Who calls this?

Reply:
The entire `SingleThreadedSynchronizationContext.cs` is copy from https://github.com/nunit/nunit/blob/master/src/NUnitFramework/framework/Internal/SingleThreadedSynchronizationContext.cs

Reply:
This sync context isn't working. It isn't single threaded, it just doesn't run the continuations. If that's the test then you could just have a noop sync context and prove that nothing ever gets scheduled on it.

Reply:
```C#
public class SingleThreadedSynchronizationContext : SynchronizationContext
{
	private readonly Queue<(SendOrPostCallback Callback, object State)> _queue = new Queue<(SendOrPostCallback Callback, object State)>();

	public override void Post(SendOrPostCallback d, object state)
	{
		_queue.Enqueue((d, state));
	}

	public override void Send(SendOrPostCallback d, object state)
	{
		base.Send(d, state);
	}

	public static void Run(Action action)
	{
		var previous = SynchronizationContext.Current;
		var context = new SingleThreadedSynchronizationContext();
		SynchronizationContext.SetSynchronizationContext(context);
		try
		{
			action();
			while (context._queue.TryDequeue(out var item))
			{
				item.Callback(item.State);
			}
		}
		finally
		{
			SynchronizationContext.SetSynchronizationContext(previous);
		}
	}
}
```

Reply:
Hi @davidfowl , It works, thanks for your help.